open_lm/model.py

```diff
@@ -117,41 +128,72 @@ def __init__(self, layer_id, args: Params):
         self.pos_embed = get_pos_embed(args)
         self.attn_fn = args.attn_func
         self.apply_qk_norm = args.apply_qk_norm
 
-        # initialize norm layers for queries and keys if needed
-        self.q_norm = (
-            args.norm_type(
-                args.n_heads * self.head_dim,
-                eps=args.norm_eps,
+        if using_te:
+                # initialize norm layers for queries and keys if needed
+            self.q_norm = (
+                te.LayerNorm(
+                    args.n_heads * self.head_dim,
+                    eps=args.norm_eps,
+                )
+                if self.apply_qk_norm
+                else nn.Identity()
             )
-            if self.apply_qk_norm
-            else nn.Identity()
-        )
-        self.k_norm = (
-            args.norm_type(
-                args.n_heads * self.head_dim,
-                eps=args.norm_eps,
+            self.k_norm = (
+                te.LayerNorm(
+                    args.n_heads * self.head_dim,
+                    eps=args.norm_eps,
+                )
+                if self.apply_qk_norm
+                else nn.Identity()
+            )
+        else:
+            # initialize norm layers for queries and keys if needed
+            self.q_norm = (
+                args.norm_type(
+                    args.n_heads * self.head_dim,
+                    eps=args.norm_eps,
+                )
+                if self.apply_qk_norm
+                else nn.Identity()
+            )
+            self.k_norm = (
+                args.norm_type(
+                    args.n_heads * self.head_dim,
+                    eps=args.norm_eps,
+                )
+                if self.apply_qk_norm
+                else nn.Identity()
             )
-            if self.apply_qk_norm
-            else nn.Identity()
-        )
 
         self.layer_id = layer_id
         self.dim = args.dim
         self.reset_parameters()
 
     def reset_parameters(self):
-        # initialize weights by trunc_normal(1/sqrt(fan_in))
-        std = 1.0 / math.sqrt(self.dim)
-        torch.nn.init.trunc_normal_(self.in_proj.weight, std=std, a=-3 * std, b=3 * std)
-        # scale init by depth as in https://arxiv.org/abs/1908.11365 -- worked slightly better.
-        std = std / math.sqrt(2 * (self.layer_id + 1))
-        torch.nn.init.trunc_normal_(self.out_proj.weight, std=std, a=-3 * std, b=3 * std)
+        if using_te:
+            # initialize weights by trunc_normal(1/sqrt(fan_in))
+            std = 1.0 / math.sqrt(self.dim)
+            torch.nn.init.trunc_normal_(self.in_proj.weight_tensor.float(), std=std, a=-3 * std, b=3 * std)
+            # scale init by depth as in https://arxiv.org/abs/1908.11365 -- worked slightly better.
+            std = std / math.sqrt(2 * (self.layer_id + 1))
+            torch.nn.init.trunc_normal_(self.out_proj.weight_tensor.float(), std=std, a=-3 * std, b=3 * std)
```

Comment:
why do we need to cast to float? does a float cast happen in place?

Reply:
We dont need float cast. Removing that in next commit.

Reply:
Removed completely as we are recursively changing NN.Linear to TE.Linear